src/ucp/core/ucp_request.inl

```diff
@@ -1003,19 +1006,29 @@ ucp_request_param_user_data(const ucp_request_param_t *param)
 
 static UCS_F_ALWAYS_INLINE ucs_memory_type_t
 ucp_request_get_memory_type(ucp_context_h context, const void *address,
-                            size_t length, const ucp_request_param_t *param)
+                            size_t count, ucp_datatype_t dt,
+                            const ucp_request_param_t *param)
 {
     ucp_memory_info_t mem_info;
 
-    if (!(param->op_attr_mask & UCP_OP_ATTR_FIELD_MEMORY_TYPE) ||
-        (param->memory_type == UCS_MEMORY_TYPE_UNKNOWN)) {
-        ucp_memory_detect(context, address, length, &mem_info);
-        ucs_assert(mem_info.type < UCS_MEMORY_TYPE_UNKNOWN);
-        return (ucs_memory_type_t)mem_info.type;
+    if ((param->op_attr_mask & UCP_OP_ATTR_FIELD_MEMORY_TYPE) &&
+        (param->memory_type != UCS_MEMORY_TYPE_UNKNOWN)) {
+        return param->memory_type;
     }
 
-    ucs_assert(param->memory_type < UCS_MEMORY_TYPE_UNKNOWN);
-    return param->memory_type;
+    if ((dt & UCP_DATATYPE_CLASS_MASK) == UCP_DATATYPE_CONTIG) {
+        ucp_memory_detect_param(context, address,
+                                ucp_contig_dt_length(dt, count), param,
```

Comment:
extra multiplication operation here

Reply:
before this change, for config datatype, we calculated ucp_contig_dt_length() once
now, for protov1, we calculate it twice - both in req->send.length init and memtype detect